File: android/src/main/java/expo/modules/rnpollardkangaroo/RnPollardKangarooModule.kt
package expo.modules.rnpollardkangaroo

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import uniffi.aptos_pollard_kangaroo_mobile.WasmKangaroo
import uniffi.aptos_pollard_kangaroo_mobile.createKangaroo

class RnPollardKangarooModule : Module() {
    private var instances: Array<WasmKangaroo> = arrayOf()

    override fun definition() = ModuleDefinition {
        // Sets the name of the module that JavaScript code will use to refer to the module. Takes a string as an argument.
        // Can be inferred from module's class name, but it's recommended to set it explicitly for clarity.
        // The module will be accessible from `requireNativeModule('RnPollardKangaroo')` in JavaScript.
        Name("RnPollardKangaroo")

        // Initialize Kangaroo Singleton
        AsyncFunction("initializeKangaroo") { secretSize: IntArray ->
            try {
                instances = Array(secretSize.size) { createKangaroo(secretSize[it].toUByte()) }
                true
            } catch (e: Exception) {
                throw Exception("Error initializing Kangaroo: ${e.localizedMessage}")
            }
        }

        // Solve DLP Function
        AsyncFunction("solveDlp") { pkBytes: ByteArray, timeLimits: LongArray ->
            try {
                for (i in instances.indices) {
                    val result = instances[i].solveDlp(pkBytes, timeLimits[i].toULong())
                    if (result != null) {
                        return@AsyncFunction result.toLong()
                    }
                }

                throw Exception("Error solving DLP: No solution found")
            } catch (e: Exception) {
                throw Exception("Error solving DLP: ${e.localizedMessage}")
            }
        }
    }
}

File: ios/RnPollardKangarooModule.swift
import ExpoModulesCore

public class RnPollardKangarooModule: Module {
    
    private var kangarooInstances: [WasmKangaroo] = []

  public func definition() -> ModuleDefinition {
    Name("RnPollardKangaroo")

      // Initialize the Kangaroo singleton
      AsyncFunction("initializeKangaroo") { (secretSizes: [UInt8]) in
          do {
              // Try createKangaroo each secretSize
            
              self.kangarooInstances = try secretSizes.map { try createKangaroo(secretSize: $0) }

              return true
          } catch {
              // Catch and throw the error
              throw NSError(domain: "RnPollardKangarooModuleError", code: 1, userInfo: [NSLocalizedDescriptionKey: error.localizedDescription])
          }
      }
      
      // Solve DLP using the initialized KangarooInstance
      AsyncFunction("solveDlp") { (pkData: Data, timeLimits: [UInt64]) -> UInt64 in
          do {
              // For each of kangarooInstances, call solveDlp with timeLimits, if result of execution is null, continue to the next execution with next time limit, if instance was the last, and no result was found - throw error
              
              for (index, kangarooInstance) in self.kangarooInstances.enumerated() {
                  if let result = try kangarooInstance.solveDlp(pk: pkData, maxTime: timeLimits[index]) {
                      return result
                  }
              }
              
              throw NSError(domain: "RnPollardKangarooModuleError", code: 2, userInfo: [NSLocalizedDescriptionKey: "No result found"])
          } catch {
              throw NSError(domain: "RnPollardKangarooModuleError", code: 3, userInfo: [NSLocalizedDescriptionKey: error.localizedDescription])
          }
      }
  }
}
